### DISCUSSION TITLE
IPC for getting information about niri

### DISCUSSION THREAD

**Discussion by EdenQwQ** [YaLTeR/niri#209] [Ideas]
It would be nice if we can have something like `niri msg activeworkspace`, making it convenient to write iteractive scripts and modules for waybar etc.

**Comment by YaLTeR** [comment]
Do you have an example from another compositor for what you're thinking?

**Reply by EdenQwQ** [reply]
hyprctl can be an example. Currently, we can only get output information from `niri msg`. It would be nice if we can also get information about workspaces, columns, windows as well.

**Comment by Supreeeme** [comment]
For workspaces in particular, that seems like it would be addressed by #135

**Comment by nebulosa2007** [comment]
Btw, is there the way to show workspaces in the waybar?

With this config have no luck: 
```
...
    "modules-left": [ ... , "wlr/workspaces"],
...
    "wlr/workspaces": {
         "format": "{name}: {icon}",
         "format-icons": {
             "active": "",
             "default": ""
         },
         "sort-by-number": true
    },

```

No errors in the status of the `niri` or `waybar` services either

**Reply by YaLTeR** [reply]
It needs https://github.com/YaLTeR/niri/issues/135. Unfortunately that protocol is not even merged into wayland-protocols yet.

**Comment by HardGraphite** [comment]
In addition to `ext-workspace-unstable` (issue #135), I think we still need tools in IPC for getting information. Niri, as a scrollable tiling compositor, provides a 2-D space. That means not only workspaces but also columns matter.

Sorry for not being familiar with most Wayland compositor, but many X11 window managers implement such tools that can be examples for reference.

In bspwm, the `bspc subscribe ...` command prints events and changes continuously.
Relevant parts in [bspwm(1)](https://man.archlinux.org/man/bspwm.1.en):

<details>

```
...

   Subscribe
       General Syntax

           subscribe [OPTIONS] (all|report|monitor|desktop|node|...)*
               Continuously print events. See the EVENTS section for the description of each
               event.

...

EVENTS
       ...

       desktop_add <monitor_id> <desktop_id> <desktop_name>
           A desktop is added.

       desktop_rename <monitor_id> <desktop_id> <old_name> <new_name>
           A desktop is renamed.

       desktop_remove <monitor_id> <desktop_id>
           A desktop is removed.

       desktop_swap <src_monitor_id> <src_desktop_id> <dst_monitor_id> <dst_desktop_id>
           A desktop is swapped.

       desktop_transfer <src_monitor_id> <src_desktop_id> <dst_monitor_id>
           A desktop is transferred.

       desktop_focus <monitor_id> <desktop_id>
           A desktop is focused.

       desktop_activate <monitor_id> <desktop_id>
           A desktop is activated.

       desktop_layout <monitor_id> <desktop_id> tiled|monocle
           The layout of a desktop changed.

       ...
```

</details>

In i3-wm, the `i3-msg` command also allows subscription.
Relevant parts in [i3-msg(1)](https://man.archlinux.org/man/i3-msg.1.en):

<details>

```
...

OPTIONS
       ...

       -t type
           Send ipc message, see below. This option defaults to "command".

       -m, --monitor
           Instead of exiting right after receiving the first subscribed event, wait indefinitely
           for all of them. Can only be used with "-t subscribe". See the "subscribe" IPC message
           type below for details.

       ...


EXAMPLES
           ...

           # Monitor window changes
           i3-msg -t subscribe -m '[ "window" ]'
```

</details>

We may have something like `niri subscribe ...` to monitor changes of workspaces and columns, so that scripts can be written to show real-time info in the status bar.

**Reply by YaLTeR** [reply]
These are good examples, thanks. I'm still considering what's the best way to implement this IPC subscription.

Would be good to also have a concrete use-case, i.e. someone wants to show this specific info, this is how it looks in another compositor, this is the script that does it. That would help inform the design of the feature, plus allow prioritizing, because otherwise adding IPC reporting to every subsystem is a lot of busywork that's not very fun on its own.

**Reply by sommerper** [reply]
I hope I'm not barging in here because my usecase is not about subscriptions - but I could write an example of what I requested in https://github.com/YaLTeR/niri/discussions/267 using hyprlands hyprctl if you're interested. I also have the bash script that accomplishes that usecase for me in Gnome - but it's rather ugly. It's basically window manipulation using info about workspaces and running apps.

**Reply by YaLTeR** [reply]
I think it's pretty clear as is, so no need
